src/jit/gentree.cpp

```diff
@@ -14938,6 +14938,13 @@ bool Compiler::gtNodeHasSideEffects(GenTree* tree, unsigned flags)
 {
     if (flags & GTF_ASG)
     {
+        // TODO-Cleanup: This only checks for GT_ASG but according to OperRequiresAsgFlag there
+        // are many more opers that are considered to have an assignment side effect: atomic ops
+        // (GT_CMPXCHG & co.), GT_MEMORYBARRIER (not classified as an atomic op) and HW intrinsic
+        // memory stores. Atomic ops have special handling in gtExtractSideEffList but the others
+        // will simply be dropped is they are ever subject to an "extract side effects" operation.
+        // It is possible that the reason no bugs have yet been observed in this area is that the
+        // other nodes are likely to always be tree roots.
```

Comment:
We probably should add tree->OperIsAtomicOp()  and GT_MEMORYBARRIER to the return true case.

This can be done in a follow up, if it desired that this be a minimal diff PR.

Reply:
I think I'll leave this for another PR. It would be good to investigate moving the atomic check from `gtExtractSideEffList` to `gtNodeHasSideEffects` as well and that can be done at the same time.